app/notifiers/notifier/base.rb

```diff
@@ -0,0 +1,79 @@
+# The base class of notifiers. This is meant to be called by the Notifications Framework
+#
+# @api notifications
+class Notifier::Base
+  class ActivityWrapper < SimpleDelegator
+    # Send notifications according to input type and recipient
+    #
+    # @param [Object] recipient The recipient of the notification
+    # @param [Symbol] type The type of notification
+    def notify(recipient, type)
+      super(recipient, type).tap do |notification|
+        case recipient
+        when Course
+          emails = @notifier.send(:notify_course, notification)
+          @pending_emails += emails unless emails.nil?
+        when User
+          email = @notifier.send(:notify_user, notification)
+          @pending_emails += email unless email.nil?
+        end
+      end
+
+      self
+    end
+
+    # Save activity and send out pending emails
+    def save(*)
+      super.tap do |result|
+        @pending_emails.each(&:deliver_later) if result
+      end
+    end
+
+    private
+
+    def initialize(notifier, activity) #:nodoc:
+      super(activity)
+      @notifier = notifier
+      @pending_emails = []
+    end
+  end
+
+  class << self
+    # This is not meant to be called by user directly
+    def method_missing(symbol, *args, &block)
+      new.send(symbol, *args, &block)
+    end
+  end
+
+  private
+
+  # Create an ActivityWrapper based on options
+  #
+  # @param [Hash] options The options used to create an activity
+  # @option options [User] :actor The actor who trigger off the activity
+  # @option options :object The object which the activity is about
+  # @option options [Symbol] :event The event name of activity
+  def create_activity(options)
+    ActivityWrapper.new(self, Activity.new(options.merge(notifier_type: self.class.name)))
+  end
+
+  # Create emails for the users of a course based on a given course notification record
+  #
+  # @param [Course::Notification] notification The notification which is used to generate emails
+  def notify_course(notification)
+    return unless notification.notification_type == 'email'
```

Comment:
@lowjoel I asked zihuan to use `if` instead of `case`. And actually I think these functions should also be renamed to `email_course` and `email_user` ( because return emails in `notify_xxx` doesn't look right), what do you think ?